db/dcrpg/pgblockchain.go

```diff
@@ -571,37 +631,117 @@ func (pgb *ChainDB) GetTicketPoolByDateAndInterval(maturityBlock int64,
+	if intervalFound && !stale {
+		// The cache was fresh.
+		return barcharts, donutCharts, height, nil
+	}
+
+	// Cache is stale. Attempt to gain updater status.
+	if !pgb.tpUpdatePermission[interval].TryLock() {
+		// Another goroutine is running db query to get the updated data.
+		if !intervalFound {
+			// Do not even have stale data. Must wait for the DB update to
+			// complete to get any data at all. Use a blocking call on the
+			// updater lock even though we are not going to actually do an
+			// update ourselves so we do not block the cache while waiting.
+			pgb.tpUpdatePermission[interval].Lock()
+			defer pgb.tpUpdatePermission[interval].Unlock()
+			// Try again to pull it from cache now that the update is completed.
+			heightSeen = pgb.Height()
+			barcharts, donutCharts, height, intervalFound, stale = TicketPoolData(interval, heightSeen)
+			// We waited for the updater of this interval, so it should be found
+			// at this point. If not, this is an error.
+			if !intervalFound {
+				log.Errorf("Charts data for interval %v failed to update.", interval)
+				return nil, nil, 0, fmt.Errorf("no charts data available")
+			}
+			if stale {
+				log.Warnf("Charts data for interval %v updated, but still stale.", interval)
+			}
+		}
+		// else return the stale data instead of waiting.
+
+		return barcharts, donutCharts, height, nil
+	}
+	// This goroutine is now the cache updater.
+	defer pgb.tpUpdatePermission[interval].Unlock()
+
+	// Retrieve chart data for best block in DB.
+	var err error
+	barcharts, donutCharts, height, err = pgb.ticketPoolVisualization(interval)
+	if err != nil {
+		log.Errorf("Failed to fetch ticket pool data: %v", err)
+		return nil, nil, 0, err
+	}
+
+	// Update the cache with the new ticket pool data.
+	UpdateTicketPoolData(interval, barcharts, donutCharts, height)
+
+	return barcharts, donutCharts, height, nil
+}
+
+// ticketPoolVisualization fetches the following ticketpool data: tickets
 // grouped on the specified interval, tickets grouped by price, and ticket
 // counts by ticket type (solo, pool, other split). The interval may be one of:
 // "mo", "wk", "day", or "all". The data is needed to populate the ticketpool
 // graphs. The data grouped by time and price are returned in a slice.
-func (pgb *ChainDB) TicketPoolVisualization(interval dbtypes.ChartGrouping) (
-	[]*dbtypes.PoolTicketsData, *dbtypes.PoolTicketsData, error) {
-	var maturityBlock = pgb.GetTicketPoolBlockMaturity()
-
-	// Tickets grouped by time interval
-	ticketsByTime, err := pgb.GetTicketPoolByDateAndInterval(maturityBlock, interval)
+func (pgb *ChainDB) ticketPoolVisualization(interval dbtypes.ChartGrouping) (byTimeAndPrice []*dbtypes.PoolTicketsData, byInputs *dbtypes.PoolTicketsData, height uint64, err error) {
+	// Ensure DB height is the same before and after queries since they are not
+	// atomic. Initial height:
+	height, err = pgb.HeightDB()
```

Comment:
You may need do away with:
```go
height, err = pgb.HeightDB()
if err != nil {
	err = fmt.Errorf("failed to query DB height: %v", err)
	return
}
```
Before the for loop, you are already sending the `height` as a variable from the function above.

Reply:
`HeightDB()` here is getting the best block processed and entered entirely into all of the DB tables.  This is the block height that you can expect to get complete charts for.

In `TicketPoolVisualization`, `heightSeen := pgb.Height()` gives the best block ever seen, but not necessarily fully processed by ChainDB.  `heightSeen` is set that way there because you should be able to get that data, although it may involve waiting (or getting stale data).

Reply:
ok. Noted

Reply:
Now that I look closer, HeightDB and Height should both indicate the block is processed in the tables of interest.  I will change to Height()